frontend/src/utils/enginePower.ts

```diff
@@ -0,0 +1,190 @@
+import { density } from "./atmosphere";
+
+export const powerFunction = (props: powerInput) => {
+  let altitudes = altitudeArrayCreator(
+    props.maxAltitude,
+    props.halfSupercharger?.endAltitude,
+    props.endSupercharger?.endAltitude,
+    props.turbocharger?.altitude
+  );
+
+  let power = Array.from(altitudes, (altitude) =>
+    powerArrayCreator(altitude, props)
+  );
+
+  let altitudesKm = altitudes.map((altitude)=>(altitude/1000)); 
+  return { x: altitudesKm, y: power };
+};
+
+interface powerInput {
+  seaLevelPower: number;
+  engineSpeed: number;
+  reductionRatio: number;
+  maxAltitude: number;
+  kCoefficient: number;
+  engineType: string;
+  halfSupercharger?: { endAltitude: number; endPower: number };
+  endSupercharger?: {
+    startAltitude: number;
+    endAltitude: number;
+    endPower: number;
+  };
+  turbocharger?: {
+    altitude: number;
+  };
+}
+
+const altitudeArrayCreator = (
+  maxAltitude: number,
+  halfSuperchargerendAltitude?: number,
+  endSuperchargerendAltitude?: number,
+  turbocharger?: number
+) => {
+  let altitudes = Array.from(
+    { length: Math.floor(1000*maxAltitude / 100) },
+    (_, i) =>  100*i
+  );
+  altitudes = [...altitudes, 1000*maxAltitude];
+
+  if (
+    halfSuperchargerendAltitude &&
+    altitudes.includes(halfSuperchargerendAltitude) === false
+  ) {
+    altitudes = [...altitudes, halfSuperchargerendAltitude];
+    altitudes.sort(function (a, b) {
+      return a - b;
+    });
+  }
+
+  if (
+    endSuperchargerendAltitude &&
+    altitudes.includes(endSuperchargerendAltitude) === false
+  ) {
+    altitudes = [...altitudes, endSuperchargerendAltitude];
+    altitudes.sort(function (a, b) {
+      return a - b;
+    });
+  }
+
+  if (turbocharger && altitudes.includes(turbocharger) === false) {
+    altitudes = [...altitudes, turbocharger];
+    altitudes.sort(function (a, b) {
+      return a - b;
+    });
+  }
+  return altitudes;
+};
+
+const powerArrayCreator = (altitude: number, props: powerInput) => {
+  let power = normalPower(altitude, 0, props.seaLevelPower, props.kCoefficient);
+
+  if (props.engineType === "piston") {
```

Comment:
I think you should switch to the array approach with the supercharger gears(speeds). There are some planes that have more than 2. It'll also be easier to reason about